File: Perez-Meneses-Juan-Carlos/Pipes/pipes.java
import java.io.IOException;
import java.io.PipedInputStream;
import java.io.PipedOutputStream;

class PipeWriter {
    private PipedInputStream inputStream;
    private PipedOutputStream outputStream;
    private String buffer;

    public PipeWriter(String buffer) {
        this.buffer = buffer;
        try {
            this.inputStream = new PipedInputStream();
            this.outputStream = new PipedOutputStream(inputStream);
        } catch (IOException e) {
            System.err.println("Error al crear la tubería: " + e.getMessage());
            System.exit(1);
        }
    }

    public void writeToPipe() {
        int i = 0;
        try {
            while (true) {
                outputStream.write(buffer.getBytes());
                System.out.print(i++ + " ");
                System.out.flush();
            }
        } catch (IOException e) {
            System.err.println("Error al escribir en la tubería: " + e.getMessage());
        }
    }

    public void closePipe() {
        try {
            inputStream.close();
            outputStream.close();
        } catch (IOException e) {
            System.err.println("Error al cerrar la tubería: " + e.getMessage());
        }
    }
}

public class Main {
    public static void main(String[] args) {
        String buffer = "a";
        PipeWriter writer = new PipeWriter(buffer);
        try {
            writer.writeToPipe();
        } finally {
            writer.closePipe();
        }
    }
}

File: Perez-Meneses-Juan-Carlos/Pipes/pipes.py
import os
import sys

class PipeWriter:
    def __init__(self, buffer):
        self.buf = buffer.encode('utf-8')  # Convertir el buffer a bytes
        self.fd_read, self.fd_write = os.pipe()  # Crear la tubería

    def write_to_pipe(self):
        i = 0
        try:
            while True:
                os.write(self.fd_write, self.buf)  # Escribir un byte en la tubería
                print(i, end=" ", flush=True)  # Mostrar el contador
                i += 1
        except KeyboardInterrupt:
            print("\nPrograma detenido manualmente.")

    def close_pipe(self):
        os.close(self.fd_read)
        os.close(self.fd_write)


if __name__ == "__main__":
    buffer = "a"
    writer = PipeWriter(buffer)
    try:
        writer.write_to_pipe()
    finally:
        writer.close_pipe()

